lib/bolt/plugin/azure.rb

```diff
@@ -0,0 +1,200 @@
+          'client_secret' => (opts['client_secret'] || config['client_secret'] || ENV['AZURE_CLIENT_SECRET']),
+          'subscription_id' => (opts['subscription_id'] || config['subscription_id'] || ENV['AZURE_SUBSCRIPTION_ID'])
+        }
+
+        # All credentials must be set, otherwise there's no point continuing
+        if creds.values.include? nil
+          keys = creds.select { |_, v| v.nil? }.keys
+          raise Bolt::ValidationError, "Missing required credentials: #{keys}"
+        end
+
+        creds
+      end
+
+      # Requests for VMs and scale sets are on a per-subscription basis
+      # You can also request VMs by a specific resource group
+      # Scale sets are always requested by resource group and scale set name
+      #
+      # Since each request only returns up to 1,000 results, requests will continue to be
+      # sent until there is no longer a nextLink token in the result set
+      def instances(token, creds, opts)
+        url = if opts['resource_group']
+                if opts['scale_set']
+                  "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+                  "resourceGroups/#{opts['resource_group']}/providers/Microsoft.Compute/" \
+                  "virtualMachineScaleSets/#{opts['scale_set']}/" \
+                  "publicipaddresses?api-version=2017-03-30"
+                else
+                  "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+                  "resourceGroups/#{opts['resource_group']}/providers/Microsoft.Network/" \
+                  "publicIPAddresses?api-version=2019-06-01"
+                end
+              else
+                "https://management.azure.com/subscriptions/#{creds['subscription_id']}/" \
+                "providers/Microsoft.Network/publicIPAddresses?api-version=2019-06-01"
+              end
+
+        header = {
+          'Authorization' => "#{token['token_type']} #{token['access_token']}"
+        }
+
+        instances = []
+
+        loop do
+          # Update the URI and make the next request
+          uri = URI.parse(url)
+          result = request(:Get, uri, nil, header)
+
+          # Add the VMs to the list of instances
+          instances += result['value']
+
+          # Continue making requests until there is no longer a nextLink
+          break unless result['nextLink']
+          url = result['nextLink']
+        end
+
+        instances
+      end
+
+      # Uses the client credentials grant flow
+      # https://docs.microsoft.com/en-us/azure/active-directory/develop/v1-oauth2-client-creds-grant-flow
+      def token(creds)
+        data = {
+          grant_type: 'client_credentials',
+          client_id: creds['client_id'],
+          client_secret: creds['client_secret'],
+          resource: 'https://management.azure.com'
+        }
+
+        uri = URI.parse("https://login.microsoftonline.com/#{creds['tenant_id']}/oauth2/token")
```

Comment:
Using an oauth workflow seems strange here. Does this require the user to create a new "application" in azure?

Reply:
Yep. It's a simple process, but creating a client application is the only way to obtain tokens to authorize with the REST API. Process is covered in the docs for the API: https://docs.microsoft.com/en-us/rest/api/azure/